backend/handler/elastic.js

```diff
@@ -0,0 +1,141 @@
+/**
+*  Copyright (C) 2021 3D Repo Ltd
+*
+*  This program is free software: you can redistribute it and/or modify
+*  it under the terms of the GNU Affero General Public License as
+*  published by the Free Software Foundation, either version 3 of the
+*  License, or (at your option) any later version.
+*
+*  This program is distributed in the hope that it will be useful,
+*  but WITHOUT ANY WARRANTY; without even the implied warranty of
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+*  GNU Affero General Public License for more details.
+*
+*  You should have received a copy of the GNU Affero General Public License
+*  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+"use strict";
+const { Client } = require("@elastic/elasticsearch");
+const logger = require("../logger");
+const systemLogger = logger.systemLogger;
+const elasticConfig = require("../config").elastic;
+const Elastic = {};
+
+const loginRecordIndex = "io-teamspace-loginrecord";
+const loginRecordMapping = {
+	"username" : { "type": "text" },
+	"loginTime" : { "type": "date" },
+	"ipAddr" : { "type": "ip" },
+	"location.country" : { "type": "text" },
+	"location.city" : { "type": "text" },
+	"referrer" : { "type": "text" },
+	"application.name": { "type": "text" },
+	"application.version": { "type": "text" },
+	"application.type": { "type": "text" },
+	"engine.name": { "type": "text" },
+	"engine.version": { "type": "text" },
+	"os.name": { "type": "text" },
+	"os.version": { "type": "text" },
+	"device": { "type": "text" }
+};
+
+const indicesMappings = [
+	{
+		index: loginRecordIndex,
+		mapping: loginRecordMapping
+	}
+];
+
+const createElasticClient = async () => {
+	if(!elasticConfig) {
+		return;
+	}
+
+	const elasticCredentials = elasticConfig.cloudAuth.split(":");
+	const config = {
+		cloud: {
+			id: elasticConfig.cloudId
+		},
+		auth: {
+			username: elasticCredentials[0],
+			password: elasticCredentials[1]
+		},
+		reload_connections: elasticConfig.reload_connections,
+		maxRetries: elasticConfig.maxRetries,
+		request_timeout: elasticConfig.request_timeout
+	};
+
+	const client = new Client(config);
+	try {
+		await client.cluster.health();
+		systemLogger.logInfo(`Succesfully connected to ${elasticConfig.cloudId.trim()}`);
+		await establishIndices(client);
+	} catch (err) {
+		systemLogger.logError("Health check failed on elastic connection, please check settings.");
+		return;
+	}
+
+	return client;
+};
+
+const establishIndices = async (elasticClient)=>{
+	indicesMappings.forEach(async (indexMapping) => {
```

Comment:
The parameter in forEach is essentially a callback function. so unless you know what you're doing it's typically not great to use async, as that promise is not waited on. i.e. as it stands, `establishIndices` will return before the indices are established.

what you want to do is actually  capture the returned promise from each iteration and wait on it, so the best thing to do here is probably `map`

```suggestion
	return Promise.all(indicesMappings.map(async (indexMapping) => {
```

Reply:
if you do this then you won't have to keep doing `indexMapping.` in the function

```suggestion
	indicesMappings.forEach(async ({index, mapping}) => {
```